shared-bindings/synthio/MidiTrack.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Artyom Skrobov
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdint.h>
+
+#include "lib/utils/context_manager_helpers.h"
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/util.h"
+#include "shared-bindings/synthio/MidiTrack.h"
+#include "supervisor/shared/translate.h"
+
+//| class MidiTrack:
+//|     """Simple square-wave MIDI synth"""
+//|
+//|     def __init__(self, buffer: ReadableBuffer, tempo: int, *, sample_rate: int = 11025) -> None:
+//|         """Create a MidiTrack from the given stream of MIDI events. Only "Note On" and "Note Off" events
```

Comment:
What do you think about not taking a `buffer` and having a `write` method instead that queues data up? That way you can pipe midi from BLE or USB directly into this object.

Reply:
I see your point, but:

* that would make looping (or even passing the same AudioSample to `play()` repeatedly) impossible;
* the protocol for `audiosample_reset_buffer_fun` has no way to report failure if the AudioSample is not rewindable;
* if the queued MIDI stream depletes during playback, the playback will stop, instead of waiting for new input. For real-time input, when the MIDI events arrive at the same speed that they're played, this would happen *after every event*!

Perhaps a new protocol, e.g. `AudioStream`, should be defined alongside `AudioSample`, and `Audio*Out` implementations will need to be updated to handle the new kind of input -- e.g. a new method `play_stream` which wouldn't take a `loop` parameter. Then `synthio` can have `MidiStream` alongside `MidiTrack`, with much of the implementation shared. Similarly, there can then be `RawStream` alongside `RawSample`.

What do you think?

Reply:
Those are all really good points. I'd like to get @jepler's thoughts on it too since he has thought about MP3 streaming some.

Reply:
@tyomitch This all looks good. Would you mind putting your comment above in an issue so we can think about that for the future? I'll merge this now.